File: Design_Patterns/BehaviouralDesignPattern/Decorator_Pattern/java/BasePizza.java
/**
 * Decorator Design Pattern - Pizza Billing System
 * Base Component: BasePizza
 * 
 * This class defines the abstract base class for all pizzas.
 * The Decorator pattern allows us to add toppings dynamically without creating
 * a separate class for every combination of pizza and toppings.
 */

public abstract class BasePizza {
    /**
     * Calculate and return the cost of the pizza.
     * This method must be implemented by all concrete pizza classes.
     * 
     * @return Cost of the pizza in rupees
     */
    public abstract int cost();
}

File: Design_Patterns/BehaviouralDesignPattern/Decorator_Pattern/python/base_pizza.py
"""
Decorator Design Pattern - Pizza Billing System
Base Component: BasePizza

This module defines the abstract base class for all pizzas.
The Decorator pattern allows us to add toppings dynamically without creating
a separate class for every combination of pizza and toppings.
"""

from abc import ABC, abstractmethod


class BasePizza(ABC):
    """
    Abstract base class for all pizzas.
    This is the Component in the Decorator pattern.
    """
    
    @abstractmethod
    def cost(self) -> int:
        """
        Calculate and return the cost of the pizza.
        
        Returns:
            int: Cost of the pizza in rupees
        """
        pass
